### PR TITLE
Add Network Contributor role assignments scoped to AKS nodepools subnets

### LINKED ISSUE #178: Azure CNI requires cluster identity to have Network Contributor permissions
According to the documentation:
https://docs.microsoft.com/en-us/azure/aks/configure-azure-cni#prerequisites

> The cluster identity used by the AKS cluster must have at least [Network Contributor](https://docs.microsoft.com/en-us/azure/role-based-access-control/built-in-roles#network-contributor) permissions on the subnet within your virtual network.

The `terraform-azurerm-aks` module by default does not take care of this, and when I tried to create a `Service` of `type: LoadBalancer` I had this issue:

```
Events:
  Type     Reason                  Age                   From                Message
  ----     ------                  ----                  ----                -------
  Normal   EnsuringLoadBalancer    4m39s (x16 over 54m)  service-controller  Ensuring load balancer
  Warning  SyncLoadBalancerFailed  4m38s (x16 over 54m)  service-controller  Error syncing load balancer: failed to ensure load balancer: Retriable: false, RetryAfter: 0s, HTTPStatusCode: 403, RawError: {"error":{"code":"AuthorizationFailed","message":"The client 'xxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx' with object id 'xxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx' does not have authorization to perform action 'Microsoft.Network/virtualNetworks/subnets/read' over scope '/subscriptions/xxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/fasthackterraform/providers/Microsoft.Network/virtualNetworks/acctvnet/subnets/subnet1' or the scope is invalid. If access was recently granted, please refresh your credentials."}}
```

This is a very well known issue and I fixed it adding to my Terraform code that calls the module the following resource:

```
# Grant AKS cluster access to use AKS subnet
resource "azurerm_role_assignment" "aks" {
  principal_id         = module.aks.system_assigned_identity[0].principal_id
  role_definition_name = "Network Contributor"
  scope                = module.network.vnet_subnets[0]
  depends_on = [module.aks]
}
```

However does it make sense to create this `azurerm_role_assignment` directly in the `terraform-azurerm-aks` module ?

### PR DESCRIPTION
**Pull request by zioproto**
This PR creates a role assignment for the AKS Service Principal to be a Network Contributor on the subnets used for the AKS Cluster.

The AKS cluster identity has the Contributor role on the AKS second resource group (MC_myResourceGroup_myAKSCluster_eastus)
In this case no additional change is necessary.

However when using a custom VNET, the AKS cluster identity needs the Network Contributor role on the VNET subnets
used by the system node pool and by any additional node pools.
Docs:
* https://learn.microsoft.com/en-us/azure/aks/configure-kubenet#prerequisites
* https://learn.microsoft.com/en-us/azure/aks/configure-azure-cni#prerequisites

This PR detects if a custom VNET is used looking at the values of the variable `var.vnet_subnet_ids` and at the value of `vnet_subnet_ids` in the map variable of the node pools.

## Issue number

Fixes #178 

## Checklist before requesting a review
- [X] The pr title can be used to describe what this pr did in `CHANGELOG.md` file
- [X] I have executed pre-commit on my machine
- [X] I have passed pr-check on my machine

### CODE REVIEW

File: locals.tf [hunk 1]
```diff
@@ -29,4 +29,12 @@ locals {
       name = var.log_analytics_workspace.name
     }
   ) : null # Finally, the Log Analytics Workspace should be disabled.
+
+  vnet_subnet_ids = toset(flatten(concat([
```
Comment (lonegunmanb) [inline locals.tf:33]: Could we rename it to `subnet_ids`?

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Thanks for opening this pr @zioproto!. Almost LGTM but only a few issues. It would be nice if we can assign this new variable in `multiple_node_pools` example.

File: variables.tf [hunk 2]
```diff
@@ -374,6 +374,12 @@ variable "cluster_name" {
   default     = null
 }
 
+variable "create_role_assignment_network_contributor" {
+  type        = bool
+  description = "Create a role assignment for the AKS Service Principal to be a Network Contributor on the subnets used for the AKS Cluster"
+  default     = false
```
[the remark this answers is not in the document]
Comment (zioproto) [reply variables.tf:381]: I added `nullable = false`, but I would like to keep the `false` default value. I dont want to introduce a change in the default behaviour.

**Comment by lonegunmanb** [comment]
Hi @zioproto this pr almost LGTM but it's base version contains a failed example that was fixed by #328, would you please rebase this pr to the latest main branch and try again? Thanks!

**Comment by zioproto** [comment]
Rebased on current `origin/main`. Thanks

**Review by lonegunmanb — APPROVED** [review]
Thanks @zioproto, LGTM! 🚀

**Pull request merged**